http/custom_roundtripper.go

```diff
@@ -0,0 +1,55 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log"
+	"net/http"
+	"time"
+
+	awsAuth "github.com/ONSdigital/dp-elasticsearch/v2/awsauth"
+)
+
+type awsSignerRoundTripper struct {
+	signer       *awsAuth.Signer
+	roundTripper http.RoundTripper
+}
+
+func NewAWSSignerRoundTripper(awsFilename, awsProfile, awsRegion, awsService string, customTransport http.RoundTripper) http.RoundTripper {
```

Comment:
And here also, any reason to return interface rather than struct?

Reply:
same as above

Reply:
https://www.ardanlabs.com/blog/2016/10/avoid-interface-pollution.html

Reply:
The article you've linked to is actually explaining why that's a bad idea :)
You can see `listing 1.` as the bad example where the returned value is an interface, in `listing 2.` (the good example) it's replaced with a struct, going on to explain 'interface pollution' as being where interfaces are used unnecessarily.

Reply:
Here if you can see my code. I am returning a struct with an interface type  which is the standard way. The article I linked explains what an interface pollution is , not that just returning an interface type can cause pollution, for interface pollution it needs to do the following
```
Here is the interface pollution smell list for the code in Listing 1:

The package declares an interface that matches the entire API of its own concrete type.
The factory function returns the interface value with the unexported concrete type value inside.
The interface can be removed and nothing changes for the user of the API.
The interface is not decoupling the API from change.
```
Just returning an interface type in the intializer won't cause a pollution. I posted the article just to make that point.

If you just look at the in built go code packages like package error, we use errors.New() , the underlying code is 

```
// because the former will succeed if err wraps an *fs.PathError.
package errors

// New returns an error that formats as the given text.
// Each call to New returns a distinct error value even if the text is identical.
func New(text string) error {
	return &errorString{text}
}
```
Here the error is an interface .

Reply:
Also please note the end of that document i posted above 
```
Use an interface:

When users of the API need to provide an implementation detail.
When API’s have multiple implementations they need to maintain internally.
When parts of the API that can change have been identified and require decoupling.
```
here the consumers can have its own implementation of the roundtripper.

Reply:
Ok, so you're not creating a new interface so it's not 'interface pollution' as per that article, but it's still a) completely unnecessary b) causing needless complication elsewhere in the code (asserting as the correct type from the interface type in your NewClient.. function) and c) is certainly not 'the standard way'. You'll actually find that 'accept interfaces return structs' is actually a well known adage and idiomatic practice in Go.

In the examples you've given, error is a special case because it's an interface that's built into the language, there's no concrete type returned from errors.New() because there is not supposed to be concrete type as a standard error.
In the http Client example the transport() function returns an interface because c.Transport is an interface (which allows for custom implementation) so it is necessary to be. The transport() function is not a constructor, it's a getter and it returns an interface because it's integral to how that specific function works - it can return many different underlying types.

In what you've done it's not at all necessary for it to be an interface as the only thing that can ever be returned is that specific type. An interface return type would be appropriate if it was possible to return more than one type that both conform to the interface.
By returning an interface nothing has been decoupled. Anything that can be done with an interface can be done by a struct that satisfies that interface. The concrete type can be accepted and passed around as the http.RoundTripper interface because it conforms, and the decoupling is done by the receiver. The user has to import this package to create the custom RoundTripper either way. From then on they can use the struct in any function that accepts a http.RoundTripper interface.
If you instead returned the concrete type no functionality would be lost form the perspective of the caller, only there would be no need to re-assert as the underlying type thus removing unnecessary complexity.

Reply:
no its not unnecessary at all, you can create multiple custom roundtrippers and wrap in like a chain and use it . So it will in the near future can be used for multiple implemenations of the roundtripper like header manipulation, logging roundtripper, tracing roundtripper etc. So it should definitely return an interface.

Reply:
I'm not sure what you mean, could you show me an example?

Reply:
for example .. in the future you could do something like 
```
func Chain2(handler roundTripHandler, roundTripper ...http.RoundTripper) Handler {
	for _, rt := range roundTripper {
		handler = rt(handler)
	}
	return handler
}
```

Reply:
in this way you can wrap your custom roundtrippers, each with its own implementation of roundtrip. If an underlying struct is returned this is not possible,

Reply:
https://github.com/googleapis/google-api-go-client/blob/main/transport/http/dial.go , example from google package 
```
// NewTransport creates an http.RoundTripper for use communicating with a Google
// cloud service, configured with the given ClientOptions. Its RoundTrip method delegates to base.
func NewTransport(ctx context.Context, base http.RoundTripper, opts ...option.ClientOption) (http.RoundTripper, error) {
	settings, err := newSettings(opts)
	if err != nil {
		return nil, err
	}
	if settings.HTTPClient != nil {
		return nil, errors.New("transport/http: WithHTTPClient passed to NewTransport")
	}
	return newTransport(ctx, base, settings)
}
```

Reply:
So I think you're wrong here - When a function accepts an interface type, you don't need to pass strictly that specific interface type. You can also pass any concrete type that satisfies that interface. The same goes for a slice of interfaces, it can be populated by multiple differing types, so long as they conform to the interface. I've just tested this with the slice implementation to be double sure but you can see here: https://go.dev/play/p/qVG5n1qU2Pl

Again in the example you've given, the reason NewTransport returns an interface is because if you look at the implementation of newTransport you can see that it can return a variety of types depending on the switch statement. If it was only possible to return one type it would return whichever concrete type that was.

Reply:
no i don't think i am wrong here . I have just send you an example with clarification

Reply:
update 1.